zeroex/orderwatch/order_watcher.go

```diff
@@ -565,6 +565,7 @@ func (w *Watcher) findOrders(makerAddress, tokenAddress common.Address, tokenID
 }
 
 func (w *Watcher) generateOrderEventsIfChanged(hashToOrderWithTxHashes map[common.Hash]*OrderWithTxHashes) error {
+	ordersColTxn := w.meshDB.Orders.OpenTransaction()
```

Comment:
As I mentioned in Slack, it's very important that we read from the database *after* opening the transaction in order to avoid the data race I described above. Otherwise, the order can get modified between when it was read from the database and when the transaction was opened. I think we can address this problem by not reading the order from the database until we are inside the `generateOrderEventsIfChanged` method.

Reply:
It's also really important to make sure that the transaction is always either committed or discarded. Right now there are some cases where we return an error before committing the transaction, and that means the orders collection would be locked forever and no writes could be made. You can fix this by adding a `defer` statement:

```go
defer func() {
    _ = ordersColTxn.Discard()
}()
```

Reply:
Good catch. Moved the transaction higher up so that it now encompasses all the reads, as well as the writes. I also added the defer.